frontend/libs/studio-components/src/components/StudioCodelistEditor/StudioCodeListEditor.tsx

```diff
@@ -0,0 +1,164 @@
+    </StudioCodeListEditorContext.Provider>
+  );
+}
+
+type InternalCodeListEditorProps = Omit<StudioCodeListEditorProps, 'texts'>;
+
+function StatefulCodeListEditor({
+  codeList: defaultCodeList,
+  onChange,
+}: InternalCodeListEditorProps): ReactElement {
+  const [codeList, setCodeList] = useState<CodeList>(defaultCodeList);
+
+  useEffect(() => {
+    setCodeList(defaultCodeList);
+  }, [defaultCodeList]);
+
+  const handleChange = useCallback(
+    (newCodeList: CodeList) => {
+      setCodeList(newCodeList);
+      onChange(newCodeList);
+    },
+    [onChange],
+  );
+
+  return <ControlledCodeListEditor codeList={codeList} onChange={handleChange} />;
+}
+
+function ControlledCodeListEditor({
+  codeList,
+  onChange,
+}: InternalCodeListEditorProps): ReactElement {
+  const { texts } = useStudioCodeListEditorContext();
+
+  const handleAddButtonClick = useCallback(() => {
+    const updatedCodeList = addEmptyCodeListItem(codeList);
+    onChange(updatedCodeList);
+  }, [codeList, onChange]);
+
+  return (
+    <fieldset className={classes.codeListEditor}>
+      <legend>{texts.codeList}</legend>
+      <CodeListTable codeList={codeList} onChange={onChange} />
+      <AddButton onClick={handleAddButtonClick} />
+    </fieldset>
+  );
+}
+
+function CodeListTable({ codeList, onChange }: InternalCodeListEditorProps): ReactElement {
+  return isCodeListEmpty(codeList) ? (
+    <EmptyCodeListTable />
+  ) : (
+    <CodeListTableWithContent codeList={codeList} onChange={onChange} />
+  );
+}
+
+function EmptyCodeListTable(): ReactElement {
+  const { texts } = useStudioCodeListEditorContext();
+  return <StudioParagraph>{texts.emptyCodeList}</StudioParagraph>;
+}
+
+function CodeListTableWithContent(props: InternalCodeListEditorProps): ReactElement {
+  return (
+    <StudioInputTable>
+      <Headings />
+      <CodeLists {...props} />
+    </StudioInputTable>
+  );
+}
+
+function Headings(): ReactElement {
+  const { texts } = useStudioCodeListEditorContext();
+
+  return (
+    <StudioInputTable.Head>
+      <StudioInputTable.Row>
+        <StudioInputTable.HeaderCell>{texts.label}</StudioInputTable.HeaderCell>
+        <StudioInputTable.HeaderCell>{texts.description}</StudioInputTable.HeaderCell>
+        <StudioInputTable.HeaderCell>{texts.value}</StudioInputTable.HeaderCell>
+        <StudioInputTable.HeaderCell>{texts.delete}</StudioInputTable.HeaderCell>
+      </StudioInputTable.Row>
+    </StudioInputTable.Head>
+  );
+}
+
+function CodeLists({ codeList, onChange }: InternalCodeListEditorProps): ReactElement {
+  const handleDeleteButtonClick = useCallback(
+    (index: number) => {
+      const updatedCodeList = removeCodeListItem(codeList, index);
+      onChange(updatedCodeList);
+    },
+    [codeList, onChange],
+  );
+
+  const handleChange = useCallback(
+    (index: number, newItem: CodeListItem) => {
+      const updatedCodeList = changeCodeListItem(codeList, index, newItem);
```

Comment:
I'm not sure exactly where we should place it, but I think we need some sort of validation on the new codelist item `value`-property here. The value cannot be duplicated. Since this is a codelist-specific component, I would expect it to validate that you cannot make a change that results in two duplicate values. 
Perhaps we could add a `validateCodelistItem` method to the utils?

Since this is a new component and not in use anywhere yet, I don't think this is a blocker for this PR necessarily. We could add validation in a separate PR.

Reply:
I totally agree. As stated in the description, I deliberately left that out of scope.

My first thought about this would be to use a similar approach as in `StudioExpression`, where the `onChange` function only is triggered when the expression is valid. This way, the user may freely edit the contents without being blocked by temporary errors, while it is only saved when the content is valid.

Reply:
Derp 🤦‍♀️ My bad!